labrad/backend.py

```diff
@@ -65,15 +68,31 @@ def sendMessage(self, target, records, *args, **kw):
 
     class TwistedConnection(BaseConnection):
         def _connect(self, password, _timeout, tls_mode):
+            @defer.inlineCallbacks
+            def _connect_deferred():
+                cxn = yield getConnection(self.host, self.port, self.name,
+                                          password, tls_mode=tls_mode)
+                self._connected.set()
+
+                @defer.inlineCallbacks
+                def await_disconnect():
+                    try:
+                        yield cxn.onDisconnect()
+                    except Exception:
+                        pass
+                    self._connected.clear()
+                await_disconnect()
```

Comment:
I've been trying to avoid explicitly adding callbacks/errbacks to Deferreds and instead using inlineCallbacks for managing the plumbing. While this _technically_ creates a new Deferred, I prefer to think of it as launching a little asynchronous coroutine in the twisted reactor. In that sense, it's not really a problem that there are no callbacks, because we don't care about the result of this coroutine, just that it clears the flag.